src/gpio.rs

```diff
@@ -546,26 +546,22 @@ macro_rules! gpio {
                         }
 
                         /// Configures the pin to operate as an open drain output pin
-                        pub fn into_open_drain_output(
-                            self,
-                            moder: &mut MODER,
-                            otyper: &mut OTYPER,
-                        ) -> $PXi<Output<OpenDrain>> {
-                            let offset = 2 * $i;
-
-                            // general purpose output mode
-                            let mode = 0b01;
-                            moder.moder().modify(|r, w| unsafe {
-                                w.bits((r.bits() & !(0b11 << offset)) | (mode << offset))
+                        pub fn into_open_drain_output(self) -> $PXi<Output<OpenDrain>> {
```

Comment:
Breaking changes in itself are fine. But changing only this, would mean other parts which take the `moder` and `otyper` register must be changed as well, to make it coherent. 

We had a discussion about this a while ago in https://github.com/stm32-rs/stm32f3xx-hal/issues/37. I've no opinion on this, but I'd rather tackle that as whole in another PR :)

Reply:
I agree, we should do this in a separate PR.

But also this implementation here is not safe. The reason we pass `&mut` references for the relevant GPIO registers in is that this way we can ensure (at compile time) that we have exclusive access to them. We need this, because otherwise race conditions can happen, as explained in #37. To make this safe we'd need something else that ensures exclusive access, like a critical section.